### PR TITLE
Add misc changes for metrics correctness

### PR DESCRIPTION
**Pull request by pmcollins**
This is in support of issue #652.

This change:
- Adds support for setting a prefix on every metric name for metrics generated for testing. Doing so will be necessary to distinguish metric names for metrics generated from different rows in a PICT file.
- Exports some fields and functions in the `correctness` package so that they can be used from the `metrics` subdirectory/package.
- Replaces argument types that refer to the struct `MockTraceConsumer` and `MockMetricsConsumer` with the interfaces `MetricsDualConsumer` and `TraceDualConsumer`, both of which can consume old and new metrics/traces. This will be used in a forthcoming `TestHarness` type that implements `MetricsDualConsumer`.
- Adds a `PrometheusDataReceiver` and a `PrometheusDataSender` so that Prometheus correctness can be tested.

### CODE REVIEW

**Comment by pmcollins** [comment]
@bogdandrutu I have added a more detailed explanation of the contents of this PR. Please let me know if anything doesn't make sense.

File: internal/goldendataset/metric_gen.go [hunk 1]
```diff
@@ -68,6 +71,18 @@ func DefaultMetricData() data.MetricData {
 }
 
 func MetricDataFromCfg(cfg MetricCfg) data.MetricData {
```
[the remark this answers is not in the document]
Comment (pmcollins) [reply internal/goldendataset/metric_gen.go:77]: Fixed

File: internal/goldendataset/metric_gen.go [hunk 2]
```diff
@@ -68,6 +71,18 @@ func DefaultMetricData() data.MetricData {
...
+	return NewMetricGenerator().GenMetricDataFromCfg(cfg)
+}
+
+type MetricGenerator struct {
```
[the remark this answers is not in the document]
Comment (pmcollins) [reply internal/goldendataset/metric_gen.go]: Fixed

File: internal/goldendataset/metric_gen.go [hunk 3]
```diff
@@ -68,6 +71,18 @@ func DefaultMetricData() data.MetricData {
...
+	metricID int
+}
+
+func NewMetricGenerator() *MetricGenerator {
```
[the remark this answers is not in the document]
Comment (pmcollins) [reply internal/goldendataset/metric_gen.go]: Fixed

File: testbed/correctness/util.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
+package correctness
```
Comment (bogdandrutu) [inline testbed/correctness/util.go:1]: Can this refactoring be split in a separate PR?

Comment (pmcollins) [reply testbed/correctness/util.go:1]: Sure

Comment (pmcollins) [reply testbed/correctness/util.go:1]: Separate PR is here: https://github.com/open-telemetry/opentelemetry-collector/pull/1497

**Comment by bogdandrutu** [comment]
Rebase?

**Comment by pmcollins** [comment]
> Rebase?

Done.

File: testbed/testbed/mock_backend.go [hunk 5]
```diff
@@ -172,13 +173,20 @@ func (mb *MockBackend) ConsumeMetricOld(md consumerdata.MetricsData) {
 	}
 }
 
+type TraceDualConsumer interface {
+	consumer.TraceConsumer
+	consumer.TraceConsumerOld
```
Comment (bogdandrutu) [inline testbed/testbed/mock_backend.go:178]: Why do you need the Old interface?

File: testbed/testbed/mock_backend.go [hunk 6]
```diff
@@ -243,29 +251,28 @@ func (tc *MockTraceConsumer) ConsumeTraceData(ctx context.Context, td consumerda
 	return nil
 }
 
+type MetricsDualConsumer interface {
+	consumer.MetricsConsumer
+	consumer.MetricsConsumerOld
```
Comment (bogdandrutu) [inline testbed/testbed/mock_backend.go:256]: Same.

File: testbed/testbed/mock_backend.go [hunk 7]
```diff
@@ -172,13 +173,20 @@ func (mb *MockBackend) ConsumeMetricOld(md consumerdata.MetricsData) {
...
 }
 
+type TraceDualConsumer interface {
+	consumer.TraceConsumer
+	consumer.TraceConsumerOld
+}
+
+var _ TraceDualConsumer = (*MockTraceConsumer)(nil)
+
 type MockTraceConsumer struct {
```
Comment (bogdandrutu) [inline testbed/testbed/mock_backend.go:183]: Does this need to be public?

File: testbed/testbed/validator.go [hunk 8]
```diff
@@ -75,437 +66,3 @@ func (v *PerfTestValidator) RecordResults(tc *TestCase) {
 		errorCause:        tc.errorCause,
 	})
 }
-
-// CorrectnessTestValidator implements TestCaseValidator for test suites using CorrectnessResults for summarizing results.
-type CorrectnessTestValidator struct {
```
Comment (bogdandrutu) [inline testbed/testbed/validator.go:80]: This extraction can be split because it is a trivial code move.

File: testbed/testbed/mock_backend.go (hunk 5 above)
Comment (pmcollins) [reply testbed/testbed/mock_backend.go:178]: To answer this and the next question, the old interface is used by Prometheus and OpenCensus on the metrics side, and Zipkin on the tracing side.

File: testbed/testbed/mock_backend.go (hunk 7 above)
Comment (pmcollins) [reply testbed/testbed/mock_backend.go:183]: Yes, it will be used by the test harness in the metrics package as a placeholder trace consumer.

File: internal/goldendataset/metric_gen.go [hunk 9]
```diff
@@ -63,11 +68,25 @@ func DefaultCfg() MetricCfg {
...
+	return newMetricGenerator().genMetricDataFromCfg(cfg)
+}
+
+type metricGenerator struct {
+	metricID int
+}
+
+func newMetricGenerator() *metricGenerator {
+	return &metricGenerator{}
+}
```
[the remark this answers is not in the document]
Comment (pmcollins) [reply internal/goldendataset/metric_gen.go:87]: `metricID` is a counter that gets incremented every time a new metric is generated. It's used to create unique metric names. I guess we could hold and increment the current count on the stack, but it would be a more complicated solution that this.

File: testbed/testbed/validator.go (hunk 8 above)
Comment (pmcollins) [reply testbed/testbed/validator.go]: Ok, I removed these changes from this PR.

**Comment by bogdandrutu** [comment]
Please rebase

**Comment by pmcollins** [comment]
> Please rebase

Done.

File: testbed/testbed/mock_backend.go (hunk 5 above)
Comment (bogdandrutu) [reply testbed/testbed/mock_backend.go:178]: That is no longer true :) for 2 out of 3 :)
